man/pysysbot.1
.TH "pysysbot" "1" "0.2.0" "Fabian Affolter" "Jabber bot for the displaying system details"
.SH "NAME"
.LP 
pysysbot \- A python based system information jabber bot
.SH "DESCRIPTION"
.LP 
This python jabber (XMPP) bot is based on the jabberbot framework
(http://thpinfo.com/2007/python-jabberbot/).  The bot is capable to
display details about the system it is running on.  Currently support
load, network configuration, system time, uptime, users who are logged
on, and kernel information.
.SH "SYNOPSIS"
.LP 
To run this program the standard way, please type:
.LP 
pysysbot
.LP 
If you run pysysbot for the first time, you be will be ask to provided your
credentials for the XMPP/Jabber account you want to use.
.LP
.SH "EXAMPLES"
.LP
Open a chat with the user of your pysysbot. With sending a keyword you will
get an answer from the bot back.
.LP
.IP "help"
This gives you an overview of the available options.
.IP "kernel"
Kernel which is used.
.IP "load"
Displays the system load over the last 1, 5, and 15 minutes.
.IP "mem"
Memory status of the system.
.IP "processes"
Shows the processes of the system.
.IP "system"
Displays details about system.
.IP "time"
Current system time.
.IP "uptime"
The uptime of the system.
.IP "version"
Details about the bot.
.LP
.SH "AUTHORS"
.LP 
Fabian Affolter <fabian@affolter-engineering.ch>


ChangeLog
pysysbot (0.0.3) Fri, 07 Jan 2011 21:49:34 +0100

  * New upstream release

 -- Fabian Affolter <fabian@affolter-engineering.ch>

pysysbot (0.0.3) Fri, 07 Jan 2011 21:48:44 +0100

  * New upstream release

 -- Fabian Affolter <fabian@affolter-engineering.ch>

pysysbot (0.0.3) Fri, 07 Jan 2011 21:38:33 +0100

  * New upstream release

 -- Fabian Affolter <fabian@affolter-engineering.ch>

pysysbot (0.0.3)

  * New upstream release

 -- Fabian Affolter <fabian@affolter-engineering.ch> Fri, 07 Jan 2011 21:25:53 +0100

commit 76e3493b95f4476393584958118ab6b4ca83d40b
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sat Jun 19 11:03:59 2010 +0200

    Updated

commit 2535fa57fc168051780472a35caabd1c981e9dcc
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sat Jun 19 10:55:24 2010 +0200

    Delete backup files from gedit

commit 744876bb7651f74c49319d31a3c5aeedfe42cf1c
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sat Jun 19 10:54:32 2010 +0200

    Minor changes

commit fe85918ccce8cc51334d3ad77a12121d407400b7
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Wed May 26 23:24:06 2010 +0200

    Conditional stuff for python added

commit 787059b4f2362da17427bc81b41d957932f9743c
Author: Fabian Affolter <fab@laptop017.local.net>
Date:   Sat Jan 30 11:17:38 2010 +0100

    Added check for installed modules

commit 0b8846e818ee3102f6fa0ba28aec22b43cf0b13e
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sun Jan 3 21:31:38 2010 +0100

    Added shebang

commit d2ad8511fbbad11a4a1fb64f606742100744b5ae
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sun Jan 3 19:09:28 2010 +0100

    Commit to sync

commit 46544312e7ebeba1924d133463c772ede9331513
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sun Jan 3 18:33:57 2010 +0100

    New layout

commit 678cec5c9e2c4647e1eb85f4acff3129fa1afe4a
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sun Jan 3 18:19:11 2010 +0100

    Added more details about memory usage

commit 343f67bf6a57edcd07f9dc613847fe1d6ae00a37
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sun Jan 3 16:00:25 2010 +0100

    Added requirement

commit 3477cbeb193b0dba85396b0b22216e828511085d
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sun Jan 3 15:49:18 2010 +0100

    Small changes

commit 2d5f352fd1850b2d3509d01164d476f854c178e2
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sun Jan 3 14:41:36 2010 +0100

    Added support for processes

commit ff776d1e884b29e99559f18b76a06c23692896b7
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sun Jan 3 00:40:15 2010 +0100

    Added pystatgrab as a requirement

commit 6ed066d8a368acf650b7f67b6cefc781fc4af4b6
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sun Jan 3 00:39:29 2010 +0100

    Updated

commit 8682c86cbd32b94256ea53f971bacdea4eac6ca3
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sat Jan 2 22:32:00 2010 +0100

    Updated

commit 5bb85c30ebbe7a5cf4367a8a4cd3875718c2f5b7
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sat Jan 2 21:59:19 2010 +0100

    Added formated load output

commit 17f830e2ffb8b7e472a3f861a61251bd9ca4311a
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sat Jan 2 19:08:21 2010 +0100

    Updated

commit 1a06bd0c6d9d67aa229884bd879c7b106572a51c
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Sat Jan 2 19:05:25 2010 +0100

    New bottom for the help message

commit ab93a5222d38d3468e86ba7bf88eebfd0dff8746
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Thu Dec 31 12:06:26 2009 +0100

    Added support for getting the ip addresses

commit c3a3000f2ec8e360cd103da398421192b223a896
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Tue Dec 29 22:20:45 2009 +0100

    Updated

commit df1bc47aeccafcda7bd3a14e1e605036a4ea12e8
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Tue Dec 29 22:13:38 2009 +0100

    Fixed name

commit 708ee4688b165aa3de307d3710129f1b8536e1b3
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Tue Dec 29 22:09:20 2009 +0100

    New directory added for cleaning

commit 9354053c1899c66b6062ca4251c190534748b3b7
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Tue Dec 29 22:06:29 2009 +0100

    setup.py updated

commit f502dfbda062197b5140afba16fed64f7f86739f
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Tue Dec 29 21:37:53 2009 +0100

    Removed man page and doc stuff not ready

commit a23796279fd30744fabe16f67734a11cefa9a88c
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Tue Dec 29 21:25:59 2009 +0100

    Fix path

commit 7c10775697bbd727f56d2fd9bc2a51420ca259da
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Tue Dec 29 21:17:55 2009 +0100

    Typo

commit ebf0a31c314ce8d0407c49003edf46e74f5fc158
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Tue Dec 29 21:13:12 2009 +0100

    Added Makefile for handling tarball and checksum cration

commit 86369112639b0dfb356d13b01126c29d8ea0762c
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Tue Dec 29 21:11:31 2009 +0100

    Updated source url

commit 6b08db738aba1ed63dd04fc75f543bf5daf49227
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Tue Dec 29 21:04:58 2009 +0100

    Initial setup.py

commit 94788865109aafade7715547ecc1a98a8cad90c5
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Tue Dec 29 21:04:06 2009 +0100

    Minor fixes

commit 122c904ad22f57b825d46a12e3717a970be9abeb
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Tue Dec 29 21:02:15 2009 +0100

    Initial spec file

commit 72a43b267fa344b1a76e16808a867231c21b2429
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Tue Dec 29 20:26:47 2009 +0100

    File moved

commit 261dd79928eaa166c9435446754f1e691f826564
Author: Fabian Affolter <fabian@bernewireless.net>
Date:   Tue Dec 29 20:25:23 2009 +0100

    Updated README

commit f73a0e8acd5fd5f25a3664b63c70995e59992b1b

pysysbot/pysysbot.py
# This file is part of pysysbot.
#
# pysysbot - A simple python jabber bot for getting system information
# Copyright (c) 2009-2015 Fabian Affolter <fabian at affolter-engineering.ch>
#
# Released under the BSD license. See COPYING file for details.
#
try:
    import datetime
    import os
    import sys
    import socket
    import ConfigParser
    import urllib2
    import psutil
    from jabberbot import JabberBot, botcmd
    import settings
except ImportError:
	print('Cannot find all required libraries please install them and try again.')
	raise SystemExit

__version__ = '0.2.0'
__author__ = 'Fabian Affolter <fabian@affolter-engineering.ch>'

class pySysBot(JabberBot):
    """
    This is a simple Jabber bot that can show you some details and
    information of the machine it is running on.
    """
    def top_of_help_message(self):
        """ Returns a string that forms the top of the help message. """
        return "pySysBot\n\n"

#Bot commands
    @botcmd
    def version(self, mess, args):
        """ Details about the bot. """
        version = '\n%s %s with Python %s\n%s\n%s' % (self.__class__.__name__,
            __version__,
            '.'.join([str(v) for v in sys.version_info[:3]]),
            'Author: %s' % __author__,
            'pysysbot is licensed under BSD'
            )
        return version

    @botcmd
    def kernel(self, mess, args):
        """Kernel which is used."""
        kernel = os.uname()[2]
        return kernel

    @botcmd
    def time(self, mess, args):
        """Current system time."""
        ctime = datetime.datetime.today().strftime('%A, %d. %B %Y %H:%M:%S')
        return ctime

    @botcmd
    def uptime(self, mess, args):
        """ The uptime of the system. """
        uptime = open('/proc/uptime').read().split()[0]
        # This is heavily based on the work of Hubert Chathi and his System status bot.
        uptime = float(uptime)
        (uptime, secs) = (int(uptime / 60), uptime % 60)
        (uptime, mins) = divmod(uptime, 60)
        (days, hours) = divmod(uptime, 24)

        uptime = '%d day%s, %d hour%s %02d min%s' % (days, days != 1 
            and 's' or '', hours, hours != 1 and 's' or '', mins,
            mins != 1 and 's' or '')
        return uptime

    @botcmd
    def system(self, mess, args):
        """ Displays details about system. """
        system = os.uname()
        system_data = "\nSystem information" + \
                "\n" + " Type:     " + system[0] + \
                "\n" + " FQDN:    " + system[1] + \
                "\n" + " Kernel:   " + system[2] + \
                "\n" + " Data:     " + system[3] + \
                "\n" + " Arch:     " + system[4]
        return system_data

    @botcmd
    def load(self, mess, args):
        """ Displays the system load over the last 1, 5, and 15 minutes. """
        loaddata = []
        load = os.getloadavg()
        for i in load:
                loaddata.append(i)
        load_data = "\nLoad average of the system" + \
                "\n" +"  1 min: \t" + str(loaddata[0]) + \
                "\n" +"  5 min: \t" + str(loaddata[1]) + \
                "\n" +" 15 min: \t" + str(loaddata[2])
        return load_data

    @botcmd
    def processes(self, mess, args):
        """ Shows the count of processes of the system. """
        processes = psutil.get_pid_list()
        return 'Running processes: %i' % (len(processes))

    @botcmd
    def disk(self, mess, args):
        # Credits: https://code.google.com/p/psutil/source/browse/examples/disk_usage.py
        """ Details about the disk usage. """
        templ = "%-35s %8s %8s %8s %5s%% %9s  %s\n"
        disks = templ % ("Device", "Total", "Used", "Free", "Use ", "Type", "Mount")
        for part in psutil.disk_partitions(all=False):
            usage = psutil.disk_usage(part.mountpoint)
            disks = disks + templ % (part.device,
                            bytes2human(usage.total),
                            bytes2human(usage.used),
                            bytes2human(usage.free),
                            int(usage.percent),
                            part.fstype,
                            part.mountpoint)
        return disks

    @botcmd
    def mem(self, mess, args):
        """ Memory status of the system. """
        vmem = psutil.virtual_memory()
        swap = psutil.swap_memory()
        mem = "\nMemory status" + \
                "\n" + " Mem total  : \t" + bytes2human(vmem[0]) + \
                "\t Swap total : \t" + bytes2human(swap[0]) + \
                "\n" + " Mem used   : \t" + bytes2human((vmem[0] - vmem[1])) + \
                "\t Swap used  : \t" + bytes2human(swap[1]) + \
                "\n" + " Mem avail. : \t" + bytes2human(vmem[1]) + \
                "\t Swap free  : \t" + bytes2human(swap[2]) + \
                "\n" + " Mem used   : \t" + str(vmem[2]) + " %" + \
                " \t Swap used  : \t" + str(swap[3]) + " %"
        return mem

    @botcmd
    def who(self, mess, args):
        """ Information about users who are currently logged in. """
        users = psutil.get_users()
        who = '\n'
        for user in users:
            who = who + '%-10s %-10s %s  (%s)\n' % \
                    (user.name,
                    user.terminal or '-',
                    datetime.datetime.fromtimestamp(user.started).strftime("%Y-%m-%d %H:%M"),
                    user.host)
        return who

def bytes2human(n):
    # Credits: http://code.activestate.com/recipes/578019
    symbols = ('K', 'M', 'G', 'T', 'P', 'E', 'Z', 'Y')
    prefix = {}
    for i, s in enumerate(symbols):
        prefix[s] = 1 << (i + 1) * 10
    for s in reversed(symbols):
        if n >= prefix[s]:
            value = float(n) / prefix[s]
            return '%.1f%s' % (value, s)
    return "%sB" % n

def main():
    config = settings.read_config('/etc/pysysbot/pysysbot.conf')
    bot = pySysBot(config[0]['username'], config[0]['password'], debug=False)
    bot.serve_forever()

if __name__ == '__main__':
    main()
